### PR TITLE
[iov-crypto] Add Xchacha20poly1305Ietf

### PR DESCRIPTION
**Pull request by webmaster128**
Xchacha20poly1305Ietf is like Chacha20poly1305Ietf bis with 192bit nonces.

See #245

### CODE REVIEW

File: packages/iov-crypto/src/libsodium.ts [hunk 1]
```diff
@@ -13,6 +13,11 @@ export type Chacha20poly1305IetfMessage = Uint8Array & As<"chacha20poly1305ietf-
 export type Chacha20poly1305IetfNonce = Uint8Array & As<"chacha20poly1305ietf-nonce">;
 export type Chacha20poly1305IetfCiphertext = Uint8Array & As<"chacha20poly1305ietf-ciphertext">;
 
+export type Xchacha20poly1305IetfKey = Uint8Array & As<"xchacha20poly1305ietf-key">;
+export type Xchacha20poly1305IetfMessage = Uint8Array & As<"xchacha20poly1305ietf-message">;
+export type Xchacha20poly1305IetfNonce = Uint8Array & As<"xchacha20poly1305ietf-nonce">;
+export type Xchacha20poly1305IetfCiphertext = Uint8Array & As<"xchacha20poly1305ietf-ciphertext">;
```
Comment (ethanfrey) [inline packages/iov-crypto/src/libsodium.ts]: Gotta love these type names ;)

File: packages/iov-crypto/src/libsodium.spec.ts [hunk 2]
```diff
@@ -621,4 +621,204 @@ describe("Libsodium", () => {
...
+        const corruptedNonce = nonce.map((x, i) => (i === 0 ? x ^ 0x01 : x)) as Xchacha20poly1305IetfNonce;
+        await Xchacha20poly1305Ietf.decrypt(ciphertext, key, corruptedNonce)
+          .then(() => fail("promise must not resolve"))
+          .catch(error => expect(error.message).toContain("invalid usage"));
+      }
+    });
+
+    it("encrypt conforms to Botan implementation ", async () => {
+      // Test data generated by
+      // echo -n "<message>" | ./botan encryption --mode=chacha20poly1305 --iv=000000000000000000000000000000000000000000000000 --ad= --key=0000000000000000000000000000000000000000000000000000000000000000 | xxd -p -c 1000000
```
Comment (ethanfrey) [inline packages/iov-crypto/src/libsodium.spec.ts]: Nice test cases.
Can you provide a link to where I can install `./botan`?

Also, why is it always with 000 for iv and key? Seems that it doesn't cover all cases?
Nevermind... Just saw that you do have random values for iv and key below. 👍

Comment (webmaster128) [reply packages/iov-crypto/src/libsodium.spec.ts]: Support for Xchacha is only available on master right now and not yet released.

```
$ git clone https://github.com/randombit/botan.git
$ cd botan
$ ./configure.py && make -j 4
$ ./botan --help
```

I created 100k test vectors to check the Botan implementation is okay ;) https://github.com/webmaster128/xchacha20poly1305-testvectors

Comment (ethanfrey) [reply packages/iov-crypto/src/libsodium.spec.ts]: Awesome. I know it all works properly, just suggesting a code comment to refer to this, if anyone comes in later to do an audit.

**Pull request merged**